### PR TITLE
Added set_extra_headers() to WebSocketServer

### PR DESCRIPTION
**Pull request by rohanrhu**
Hi, I'm resending PR to `master` branch with changes.

### CODE REVIEW

**Comment by rohanrhu** [comment]
Please merge this PR. I'm still using my revision on my server because I need sending CORS headers for my game to connect my server from itch.io.

File: modules/websocket/doc_classes/WebSocketServer.xml [hunk 1]
```diff
@@ -58,6 +58,15 @@
 				Returns [code]true[/code] if the server is actively listening on a port.
 			</description>
 		</method>
+		<method name="set_extra_headers">
+			<return type="void">
+			</return>
+			<argument index="0" name="headers" type="PackedStringArray" default="PackedStringArray(  )">
+			</argument>
+			<description>
+				Sets extra headers.
```
Comment (Faless) [inline modules/websocket/doc_classes/WebSocketServer.xml:67]:
```suggestion
				Sets additional headers to be sent to clients during the HTTP handshake.
```

File: modules/websocket/emws_server.cpp [hunk 2]
```diff
@@ -33,6 +33,9 @@
 #include "emws_server.h"
 #include "core/os/os.h"
 
+void EMWSServer::set_extra_headers(Vector<String> headers) {
```
Comment (Faless) [inline modules/websocket/emws_server.cpp:36]:
```suggestion
void EMWSServer::set_extra_headers(const Vector<String> &p_headers) {
```

File: modules/websocket/emws_server.h [hunk 3]
```diff
@@ -42,6 +42,7 @@ class EMWSServer : public WebSocketServer {
 
 public:
 	Error set_buffers(int p_in_buffer, int p_in_packets, int p_out_buffer, int p_out_packets);
+	void set_extra_headers(Vector<String> headers);
```
Comment (Faless) [inline modules/websocket/emws_server.h:45]:
```suggestion
	void set_extra_headers(const Vector<String> &p_headers);
```

File: modules/websocket/websocket_server.h [hunk 4]
```diff
@@ -50,6 +50,7 @@ class WebSocketServer : public WebSocketMultiplayerPeer {
 	Ref<X509Certificate> ca_chain;
 
 public:
+	virtual void set_extra_headers(Vector<String> headers) = 0;
```
Comment (Faless) [inline modules/websocket/websocket_server.h:53]:
```suggestion
	virtual void set_extra_headers(const Vector<String> &p_headers) = 0;
```

File: modules/websocket/wsl_server.cpp [hunk 5]
```diff
@@ -104,7 +104,7 @@ bool WSLServer::PendingPeer::_parse_request(const Vector<String> p_protocols) {
 	return true;
 }
 
-Error WSLServer::PendingPeer::do_handshake(const Vector<String> p_protocols) {
+Error WSLServer::PendingPeer::do_handshake(const Vector<String> p_protocols, Vector<String> &extra_headers) {
```
Comment (Faless) [inline modules/websocket/wsl_server.cpp:107]:
```suggestion
Error WSLServer::PendingPeer::do_handshake(const Vector<String> p_protocols, const Vector<String> &p_extra_headers) {
```

File: modules/websocket/wsl_server.cpp [hunk 6]
```diff
@@ -166,6 +169,10 @@ Error WSLServer::PendingPeer::do_handshake(const Vector<String> p_protocols) {
 	return OK;
 }
 
+void WSLServer::set_extra_headers(Vector<String> headers) {
```
Comment (Faless) [inline modules/websocket/wsl_server.cpp:172]:
```suggestion
void WSLServer::set_extra_headers(const Vector<String> &p_headers) {
```

File: modules/websocket/wsl_server.h [hunk 7]
```diff
@@ -66,7 +66,7 @@ class WSLServer : public WebSocketServer {
 
 		PendingPeer();
 
-		Error do_handshake(const Vector<String> p_protocols);
+		Error do_handshake(const Vector<String> p_protocols, Vector<String> &extra_headers);
```
Comment (Faless) [inline modules/websocket/wsl_server.h:69]:
```suggestion
		Error do_handshake(const Vector<String> p_protocols, const Vector<String> &p_extra_headers);
```

File: modules/websocket/wsl_server.h [hunk 8]
```diff
@@ -77,9 +77,11 @@ class WSLServer : public WebSocketServer {
 	List<Ref<PendingPeer>> _pending;
 	Ref<TCP_Server> _server;
 	Vector<String> _protocols;
+	Vector<String> _extra_headers;
 
 public:
 	Error set_buffers(int p_in_buffer, int p_in_packets, int p_out_buffer, int p_out_packets);
+	void set_extra_headers(Vector<String> headers);
```
Comment (Faless) [inline modules/websocket/wsl_server.h:84]:
```suggestion
	void set_extra_headers(const Vector<String> &p_headers);
```

**Review by Faless — CHANGES_REQUESTED** [review]
See comments (`const` usage and `p_` parameters prefix).
Looks overall good but needs a rebase :+1: .

**Comment by Faless** [comment]
> because I need sending CORS headers for my game to connect my server from itch.io.

While the PR scope is okay, I **strongly advise** (**STRONGLY!**) to put an HTTP proxy in front of it to manage both the extra headers and **requests sanitization**, [e.g. with nginx](https://nginx.org/en/docs/http/websocket.html)

**Comment by rohanrhu** [comment]
> > because I need sending CORS headers for my game to connect my server from itch.io.
> 
> While the PR scope is okay, I **strongly advise** (**STRONGLY!**) to put an HTTP proxy in front of it to manage both the extra headers and **requests sanitization**, [e.g. with nginx](https://nginx.org/en/docs/http/websocket.html)

You mean using proxy for just sending additional headers? We can just send CORS headers when it is neccessary, why would we need to use a proxy?

**Comment by rohanrhu** [comment]
Should I resolve conflicts or you do that?

**Comment by Faless** [comment]
> Should I resolve conflicts or you do that?

I can't resolve the conflicts on your branch, please rebase and squash the commits, see here for instructions:
https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase

**Comment by rohanrhu** [comment]
> > Should I resolve conflicts or you do that?
> 
> I can't resolve the conflicts on your branch, please rebase and squash the commits, see here for instructions: https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase

Thank you @Faless I will resolve conflicts soon.

**Comment by Faless** [comment]
> We can just send CORS headers when it is neccessary, why would we need to use a proxy?

As a bare minimum (it can offer more than that):
- Access/Error logging
- Requests filtering

I.e. a basic form of protection against the dozens of malicious requests (port scanners/exploit checkers/etc) you'll receive every day and that will otherwise end up being consumed by godot (most likely at a higher cost, and resulting in more errors in the godot log unrelated to the actual game).

**Comment by rohanrhu** [comment]
Hiiiii @Faless I resolved conflicts with latest revision.

File: modules/websocket/emws_server.h [hunk 9]
```diff
@@ -42,6 +42,7 @@ class EMWSServer : public WebSocketServer {
 
 public:
 	Error set_buffers(int p_in_buffer, int p_in_packets, int p_out_buffer, int p_out_packets) override;
+	void set_extra_headers(const Vector<String> &p_headers);
```
Comment (Faless) [inline modules/websocket/emws_server.h:45]: This needs to be marked as override
```suggestion
	void set_extra_headers(const Vector<String> &p_headers) override;
```

**Review by Faless — CHANGES_REQUESTED** [review]
Good job :+1: 
See review comment on adding the `override` keyword to `set_extra_headers` when overriding.
The commits also needs to be squashed together https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

File: modules/websocket/wsl_server.h [hunk 10]
```diff
@@ -73,9 +73,11 @@ class WSLServer : public WebSocketServer {
 	List<Ref<PendingPeer>> _pending;
 	Ref<TCPServer> _server;
 	Vector<String> _protocols;
+	Vector<String> _extra_headers;
 
 public:
 	Error set_buffers(int p_in_buffer, int p_in_packets, int p_out_buffer, int p_out_packets) override;
+	void set_extra_headers(const Vector<String> &p_headers);
```
Comment (Faless) [inline modules/websocket/wsl_server.h:80]:
```suggestion
	void set_extra_headers(const Vector<String> &p_headers) override;
```

**Comment by rohanrhu** [comment]
Hiiii @Faless I squashed it into one commit.

**Comment by Faless** [comment]
CI is failing: https://github.com/godotengine/godot/runs/5964150172?check_suite_focus=true .
I think you just need to [re-update the class reference](https://docs.godotengine.org/en/stable/community/contributing/updating_the_class_reference.html#updating-the-documentation-template) (`godot --doctool
`) and re-squash.

**Comment by rohanrhu** [comment]
> CI is failing: https://github.com/godotengine/godot/runs/5964150172?check_suite_focus=true . I think you just need to [re-update the class reference](https://docs.godotengine.org/en/stable/community/contributing/updating_the_class_reference.html#updating-the-documentation-template) (`godot --doctool `) and re-squash.

Meow.. I did it.

**Comment by Faless** [comment]
I think CI will fail again, the doctool seem to not have worked properly :crying_cat_face: .
Can you please try applying this patch?

```diff
diff --git a/modules/websocket/doc_classes/WebSocketServer.xml b/modules/websocket/doc_classes/WebSocketServer.xml
index ef3279aac4..46b0274de3 100644
--- a/modules/websocket/doc_classes/WebSocketServer.xml
+++ b/modules/websocket/doc_classes/WebSocketServer.xml
@@ -60,6 +60,13 @@
 				If [code]false[/code] is passed instead (default), you must call [PacketPeer] functions ([code]put_packet[/code], [code]get_packet[/code], etc.), on the [WebSocketPeer] returned via [code]get_peer(id)[/code] to communicate with the peer with given [code]id[/code] (e.g. [code]get_peer(id).get_available_packet_count[/code]).
 			</description>
 		</method>
+		<method name="set_extra_headers">
+			<return type="void" />
+			<argument index="0" name="headers" type="PackedStringArray" default="PackedStringArray()" />
+			<description>
+				Sets additional headers to be sent to clients during the HTTP handshake.
+			</description>
+		</method>
 		<method name="stop">
 			<return type="void" />
 			<description>
```

**Review by akien-mga — APPROVED** [review]

**Review by Faless — APPROVED** [review]

**Comment by Faless** [comment]
Thanks! :cat2: :trophy:

**Comment by akien-mga** [comment]
Cherry-picked for 3.5.

**Pull request merged**